File: android/app/src/main/java/com/vocahq/vocaphone/core/UsageStats.kt
package com.vocahq.vocaphone.core

import java.text.BreakIterator
import java.time.Instant
import java.time.LocalDate
import java.util.Calendar
import java.util.Locale
import java.util.TimeZone
import org.json.JSONObject

/**
 * On-device dictation counters: how much you have dictated, how fast you speak,
 * and how many days in a row you have used it.
 *
 * Counts only. No transcript text, no audio, no gateway, no reporting — the
 * whole type is a handful of numbers plus a bounded map of day totals, and it
 * never leaves the phone.
 *
 * All arithmetic lives here rather than in the repository so that every rule
 * worth checking — word counting, streaks, pruning, lenient decoding — is a pure
 * function with a test, the way [Snippet] and the rest of `core/` are written.
 */
data class UsageStats(
    val totalWords: Long = 0,
    val totalTranscriptions: Long = 0,
    val totalAudioMillis: Long = 0,
    val lastDayKey: String = "",
    val currentStreak: Int = 0,
    val bestStreak: Int = 0,
    val dailyWords: Map<String, Int> = emptyMap(),
) {

    val averageWordsPerMinute: Double
        get() = if (totalAudioMillis <= 0) 0.0 else totalWords / (totalAudioMillis / 60_000.0)

    val hasAny: Boolean get() = totalTranscriptions > 0

    fun currentStreakAt(now: Long): Int {
        if (lastDayKey.isEmpty()) return 0
        val elapsed = daysBetween(lastDayKey, dayKey(now)) ?: return 0
        return when {
            isUnreconcilableFuture(elapsed) -> 0
            elapsed <= 1 -> currentStreak
            else -> 0
        }
    }

    fun record(transcript: String, durationMillis: Long?, now: Long): UsageStats {
        val words = wordCount(transcript)
        if (words == 0) return this

        val key = dayKey(now)
        val daily = dailyWords.toMutableMap()
        daily[key] = (daily[key] ?: 0) + words

        val streak = advanceStreak(currentStreak, lastDayKey, key)
        return copy(
            totalWords = totalWords + words,
            totalTranscriptions = totalTranscriptions + 1,
            totalAudioMillis = totalAudioMillis + (durationMillis?.coerceAtLeast(0) ?: 0),
            lastDayKey = nextDayKey(lastDayKey, key),
            currentStreak = streak,
            bestStreak = maxOf(bestStreak, streak),
            dailyWords = pruneDaily(daily),
        )
    }

    companion object {
        const val DAILY_LIMIT = 7

        fun wordCount(text: String): Int {
            val trimmed = text.trim()
            if (trimmed.isEmpty()) return 0
            val iterator = BreakIterator.getWordInstance(Locale.ROOT)
            iterator.setText(trimmed)
            var count = 0
            var start = iterator.first()
            var end = iterator.next()
            while (end != BreakIterator.DONE) {
                val segment = trimmed.substring(start, end)
                if (segment.any(Char::isLetterOrDigit)) count++
                start = end
                end = iterator.next()
            }
            return count
        }

        fun dayKey(millis: Long, zone: TimeZone = TimeZone.getDefault()): String {
            val calendar = Calendar.getInstance(zone, Locale.ROOT)
            calendar.timeInMillis = millis
            return String.format(
                Locale.ROOT,
                "%04d-%02d-%02d",
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH) + 1,
                calendar.get(Calendar.DAY_OF_MONTH),
            )
        }

        /**
         * Milliseconds from [fromMillis] until the start of the next local day.
         *
         * The mirror of [dayKey]: that says which day an instant falls in, this
         * says when that day ends, so a caller that sleeps for this long wakes
         * on the first instant [dayKey] labels differently.
         *
         * Uses `atStartOfDay` rather than a [Calendar] rolled back to midnight,
         * because midnight is not always a single instant. Havana and the
         * Azores change their clocks *at* midnight, so local 00:00 happens
         * twice that night and [Calendar] resolves the ambiguity to the later
         * one — an hour after the day has already turned. `atStartOfDay` is
         * specified to take the earlier offset in an overlap, and the instant
         * after the gap on a night where 00:00 never happens at all.
         *
         * Floored at one second: a device whose clock moves while this is being
         * read could otherwise return zero, and a caller that reschedules
         * itself on the result would spin.
         */
        fun millisUntilNextDay(
            fromMillis: Long,
            zone: TimeZone = TimeZone.getDefault(),
        ): Long {
            val id = zone.toZoneId()
            val nextMidnight = Instant.ofEpochMilli(fromMillis)
                .atZone(id)
                .toLocalDate()
                .plusDays(1)
                .atStartOfDay(id)
                .toInstant()
                .toEpochMilli()
            return (nextMidnight - fromMillis).coerceAtLeast(1_000L)
        }

        fun daysBetween(fromKey: String, toKey: String): Int? = runCatching {
            (LocalDate.parse(toKey).toEpochDay() - LocalDate.parse(fromKey).toEpochDay()).toInt()
        }.getOrNull()

        fun isDayKey(value: String): Boolean =
            value.isNotEmpty() && runCatching { LocalDate.parse(value) }.isSuccess

        fun advanceStreak(current: Int, lastDayKey: String, todayKey: String): Int {
            if (lastDayKey.isEmpty()) return 1
            val elapsed = daysBetween(lastDayKey, todayKey) ?: return 1
            return when {
                isUnreconcilableFuture(elapsed) -> 1
                elapsed <= 0 -> current.coerceAtLeast(1)
                elapsed == 1 -> current.coerceAtLeast(0) + 1
                else -> 1
            }
        }

        /** The stored day after recording on [todayKey], repaired if it cannot be trusted. */
        fun nextDayKey(stored: String, todayKey: String): String {
            if (!isDayKey(stored)) return todayKey
            val elapsed = daysBetween(stored, todayKey)
            if (elapsed != null && isUnreconcilableFuture(elapsed)) return todayKey
            return maxOf(stored, todayKey)
        }

        /**
         * Whether a stored day sits so far ahead of today that it is a wrong
         * clock rather than a streak.
         *
         * A stored day can legitimately be ahead: correcting a clock backwards
         * leaves one there, and a run should survive that — which is why a small
         * negative gap holds the streak rather than breaking it. But a day
         * further ahead than the entire retained window cannot be reconciled with
         * anything the reader can see, and left alone it never expires and never
         * advances, because every later comparison stays negative. A device that
         * once recorded a dictation dated 2099 would otherwise be stuck for good.
         *
         * The boundary is [DAILY_LIMIT] because that is the history the screen
         * can actually show; past it there is nothing to reconcile against.
         */
        private fun isUnreconcilableFuture(elapsed: Int): Boolean = elapsed < -DAILY_LIMIT

        /**
         * The [DAILY_LIMIT] most recent days by date, not by insertion order.
         *
         * Readable labels are ranked first. Sorting on the string alone would
         * put an unreadable key above every real date — "zzz" beats "2026-09-09"
         * — so with the window full a damaged label would survive and evict a
         * genuine day's word count. The map is deliberately not validated when it
         * is decoded, precisely so that a bad label costs nothing; it must not
         * cost a day here instead.
         */
        fun pruneDaily(daily: Map<String, Int>): Map<String, Int> {
            if (daily.size <= DAILY_LIMIT) return daily.toMap()
            return daily.entries
                .sortedWith(
                    compareByDescending<Map.Entry<String, Int>> { isDayKey(it.key) }
                        .thenByDescending { it.key },
                )
                .take(DAILY_LIMIT)
                .associate { it.key to it.value }
        }

        fun encode(stats: UsageStats): String = JSONObject().apply {
            put("totalWords", stats.totalWords)
            put("totalTranscriptions", stats.totalTranscriptions)
            put("totalAudioMillis", stats.totalAudioMillis)
            put("lastDayKey", stats.lastDayKey)
            put("currentStreak", stats.currentStreak)
            put("bestStreak", stats.bestStreak)
            put(
                "daily",
                JSONObject().apply {
                    stats.dailyWords.forEach { (day, words) -> put(day, words) }
                },
            )
        }.toString()

        /**
         * Field by field with defaults, so a payload written by an older or
         * newer build still yields every value it does carry.
         *
         * A whole-object decode that threw would reset somebody's lifetime
         * totals the first time a field was added, which is the one failure this
         * type must not have. The day map is decoded separately for the same
         * reason: a malformed map costs the activity list, not the totals.
         */
        fun decode(stored: String?): UsageStats {
            if (stored.isNullOrBlank()) return UsageStats()
            return runCatching {
                val json = JSONObject(stored)
                val daily = decodeDaily(json.optJSONObject("daily"))
                UsageStats(
                    totalWords = json.optLong("totalWords", 0).coerceAtLeast(0),
                    totalTranscriptions = json.optLong("totalTranscriptions", 0).coerceAtLeast(0),
                    totalAudioMillis = json.optLong("totalAudioMillis", 0).coerceAtLeast(0),
                    // The stored day if it is still readable, else the newest day
                    // the retained map can vouch for — which also covers a
                    // payload written before this field existed.
                    //
                    // Both sources are checked, not just the first. The day map
                    // is deliberately not validated when it is decoded, so its
                    // largest key can itself be unreadable; taking `maxOrNull`
                    // blindly would let the recovery hand back the very kind of
                    // value it is recovering from. Filtering also means one bad
                    // key does not cost the good days beside it.
                    lastDayKey = json.optString("lastDayKey").takeIf(::isDayKey)
                        ?: daily.keys.filter(::isDayKey).maxOrNull().orEmpty(),
                    currentStreak = json.optInt("currentStreak", 0).coerceAtLeast(0),
                    bestStreak = json.optInt("bestStreak", 0).coerceAtLeast(0),
                    dailyWords = daily,
                )
            }.getOrDefault(UsageStats())
        }

        private fun decodeDaily(json: JSONObject?): Map<String, Int> {
            if (json == null) return emptyMap()
            return runCatching {
                buildMap {
                    json.keys().forEach { key ->
                        val words = json.optInt(key, 0)
                        if (words > 0) put(key, words)
                    }
                }.let(::pruneDaily)
            }.getOrDefault(emptyMap())
        }

    }
}

File: ios/VocaPhoneShared/UsageStats.swift
import Foundation

struct UsageStats: Codable, Equatable, Sendable {
    var totalWords: Int = 0
    var totalDictations: Int = 0
    var totalSeconds: Double = 0
    var lastDayKey: String = ""
    var currentStreak: Int = 0
    var bestStreak: Int = 0
    var dailyWords: [String: Int] = [:]
    var dailyDictations: [String: Int] = [:]
    var dailySeconds: [String: Double] = [:]

    static let dailyLimit = 7

    var hasAny: Bool { totalDictations > 0 }

    var averageWordsPerMinute: Double {
        guard totalSeconds > 0 else { return 0 }
        return Double(totalWords) / (totalSeconds / 60)
    }

    func currentStreak(at now: Date, timeZone: TimeZone = .current) -> Int {
        guard !lastDayKey.isEmpty else { return 0 }
        guard let elapsed = Self.daysBetween(lastDayKey, Self.dayKey(now, timeZone: timeZone)) else {
            return 0
        }
        if Self.isUnreconcilableFuture(elapsed) { return 0 }
        return elapsed <= 1 ? currentStreak : 0
    }

    func recording(dayKey key: String, words: Int, seconds: Double) -> UsageStats {
        guard words > 0 else { return self }
        var next = self
        next.totalWords += words
        next.totalDictations += 1
        next.totalSeconds += max(seconds, 0)
        var daily = dailyWords
        daily[key, default: 0] += words
        var dictations = dailyDictations
        dictations[key, default: 0] += 1
        var secondsByDay = dailySeconds
        secondsByDay[key, default: 0] += max(seconds, 0)
        let retained = Set(Self.rankedDayKeys(in: daily).prefix(Self.dailyLimit))
        next.dailyWords = daily.filter { retained.contains($0.key) }
        next.dailyDictations = dictations.filter { retained.contains($0.key) }
        next.dailySeconds = secondsByDay.filter { retained.contains($0.key) }
        let streak = Self.advanceStreak(currentStreak, lastDayKey: lastDayKey, todayKey: key)
        next.currentStreak = streak
        next.bestStreak = max(bestStreak, streak)
        next.lastDayKey = Self.nextDayKey(lastDayKey, todayKey: key)
        return next
    }

    private static func gregorian(_ timeZone: TimeZone) -> Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        return calendar
    }

    static func dayKey(_ date: Date, timeZone: TimeZone = .current) -> String {
        let parts = gregorian(timeZone).dateComponents([.year, .month, .day], from: date)
        return String(
            format: "%04d-%02d-%02d",
            parts.year ?? 0,
            parts.month ?? 0,
            parts.day ?? 0
        )
    }

    static func daysBetween(_ fromKey: String, _ toKey: String) -> Int? {
        guard let from = epochDay(fromKey), let to = epochDay(toKey) else { return nil }
        return to - from
    }

    static func isDayKey(_ value: String) -> Bool { epochDay(value) != nil }

    private static func epochDay(_ key: String) -> Int? {
        let parts = key.split(separator: "-", omittingEmptySubsequences: false)
        guard key.count == 10, parts.count == 3,
              let year = Int(parts[0]), let month = Int(parts[1]), let day = Int(parts[2]),
              parts[0].count == 4, parts[1].count == 2, parts[2].count == 2
        else { return nil }
        let calendar = gregorian(TimeZone(secondsFromGMT: 0) ?? .current)
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = day
        guard let date = calendar.date(from: components) else { return nil }
        let back = calendar.dateComponents([.year, .month, .day], from: date)
        guard back.year == year, back.month == month, back.day == day else { return nil }
        return Int((date.timeIntervalSince1970 / 86_400).rounded(.down))
    }

    static func advanceStreak(_ current: Int, lastDayKey: String, todayKey: String) -> Int {
        guard !lastDayKey.isEmpty else { return 1 }
        guard let elapsed = daysBetween(lastDayKey, todayKey) else { return 1 }
        if isUnreconcilableFuture(elapsed) { return 1 }
        if elapsed <= 0 { return max(current, 1) }
        if elapsed == 1 { return max(current, 0) + 1 }
        return 1
    }

    static func nextDayKey(_ stored: String, todayKey: String) -> String {
        guard isDayKey(stored) else { return todayKey }
        if let elapsed = daysBetween(stored, todayKey), isUnreconcilableFuture(elapsed) {
            return todayKey
        }
        return max(stored, todayKey)
    }

    private static func isUnreconcilableFuture(_ elapsed: Int) -> Bool {
        elapsed < -dailyLimit
    }

    static func pruneDaily(_ daily: [String: Int]) -> [String: Int] {
        guard daily.count > dailyLimit else { return daily }
        let ranked = rankedDayKeys(in: daily)
        return Dictionary(uniqueKeysWithValues: ranked.prefix(dailyLimit).map { ($0, daily[$0] ?? 0) })
    }

    private static func rankedDayKeys(in daily: [String: Int]) -> [String] {
        daily.keys.sorted { left, right in
            let leftReadable = isDayKey(left)
            let rightReadable = isDayKey(right)
            if leftReadable != rightReadable { return leftReadable }
            return left > right
        }
    }

    func recentDays(limit: Int = dailyLimit) -> [DailyUsage] {
        dailyWords
            .filter { Self.isDayKey($0.key) }
            .sorted { $0.key > $1.key }
            .prefix(limit)
            .map {
                DailyUsage(
                    key: $0.key,
                    words: $0.value,
                    dictations: dailyDictations[$0.key, default: 0],
                    seconds: dailySeconds[$0.key, default: 0]
                )
            }
    }

    /// Seven consecutive calendar days, including quiet days, for a chart that
    /// shows an actual trend rather than only the dates that have activity.
    func lastSevenDays(endingAt now: Date, timeZone: TimeZone = .current) -> [DailyUsage] {
        let calendar = Self.gregorian(timeZone)
        let today = calendar.startOfDay(for: now)
        return (0..<Self.dailyLimit).reversed().compactMap { offset in
            guard let date = calendar.date(byAdding: .day, value: -offset, to: today) else {
                return nil
            }
            let key = Self.dayKey(date, timeZone: timeZone)
            return DailyUsage(
                key: key,
                words: dailyWords[key, default: 0],
                dictations: dailyDictations[key, default: 0],
                seconds: dailySeconds[key, default: 0]
            )
        }
    }

    static func wordCount(_ text: String) -> Int {
        var count = 0
        text.enumerateSubstrings(
            in: text.startIndex..<text.endIndex,
            options: [.byWords, .localized]
        ) { _, _, _, _ in
            count += 1
        }
        return count
    }

    static func nextDayStart(after date: Date, timeZone: TimeZone = .current) -> Date {
        let calendar = gregorian(timeZone)
        let today = calendar.startOfDay(for: date)
        guard let tomorrow = calendar.date(byAdding: .day, value: 1, to: today) else {
            return date.addingTimeInterval(86_400)
        }
        return calendar.startOfDay(for: tomorrow)
    }
}

struct DailyUsage: Equatable, Identifiable, Sendable {
    var id: String { key }
    let key: String
    let words: Int
    let dictations: Int
    let seconds: Double
}

extension UsageStats {
    /// New daily dimensions decode leniently so a build upgrade preserves a
    /// summary written before sessions and voice time were tracked per day.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        totalWords = try container.decodeIfPresent(Int.self, forKey: .totalWords) ?? 0
        totalDictations = try container.decodeIfPresent(Int.self, forKey: .totalDictations) ?? 0
        totalSeconds = try container.decodeIfPresent(Double.self, forKey: .totalSeconds) ?? 0
        lastDayKey = try container.decodeIfPresent(String.self, forKey: .lastDayKey) ?? ""
        currentStreak = try container.decodeIfPresent(Int.self, forKey: .currentStreak) ?? 0
        bestStreak = try container.decodeIfPresent(Int.self, forKey: .bestStreak) ?? 0
        dailyWords = try container.decodeIfPresent([String: Int].self, forKey: .dailyWords) ?? [:]
        dailyDictations = try container.decodeIfPresent(
            [String: Int].self,
            forKey: .dailyDictations
        ) ?? [:]
        dailySeconds = try container.decodeIfPresent(
            [String: Double].self,
            forKey: .dailySeconds
        ) ?? [:]
    }
}
